### PR TITLE
Pull request template

### PR DESCRIPTION
**Pull request by Beladric**
New pull request templates for React template and Python template are available. They look like each other, with only one difference. 

I would like to know If I create one template per project and more changes between them will arrive later, or if I create only one global template for all projects ?

### CODE REVIEW

**Comment by galyfray** [comment]
I think you misunderstood the meaning of "default templates". Default templates are generic, they are shared by all the repository that will be created in the Libra foundation. If a repository need more specific templates, we can override them by creating a template in the said repository. This mean that creating here a template for the React repo and another for the python repo does not make sens.

File: PULL_REQUEST_TEMPLATE/PYTHON_TEMPLATE_PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+- [ ] Any dependent changes have been merged and published in downstream modules
+
+Your pull request need to hold following parts.
+
+## Description
+
+Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List
+any dependencies that are required for this change.
+
+## Type of change
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PYTHON_TEMPLATE_PULL_REQUEST_TEMPLATE.md:24]: The type of change could be signaled through labels or a tag at the beginning of the PR's title. Either of these ways would enable us to see the type of change before reading the PR.

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+## Prerequisites
+
+***This part help you to don't forget anything before create pull request, If everything seem good you could delete it***
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:3]:
```suggestion
***This part help you to don't forget anything before create pull request, If everything seem good you can delete it***
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+Thanks for contribution :heart:. Before creating pull
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:5]:
```suggestion
Thanks for your contribution :heart:. Before creating pull
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+request you were able to check :
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:6]:
```suggestion
requestplease make sure that:
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+- [ ] Your code follows the style guidelines of this project
+- [ ] You have performed a self-review of my code
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:9]:
```suggestion
- [ ] You have performed a self-review of your code
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+- [ ] You have commented my code, particularly in hard-to-understand areas
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:10]:
```suggestion
- [ ] You have commented your code, particularly in hard-to-understand areas
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+- [ ] You have made corresponding changes to the documentation
+- [ ] Your changes generate no new warnings
+- [ ] You have added tests that prove my fix is effective or that my feature works
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:13]:
```suggestion
- [ ] You have added tests that prove your fix is effective or that your feature works
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+- [ ] New and existing unit tests pass locally with your changes
+- [ ] Any dependent changes have been merged and published in downstream modules
+
+Your pull request title must contain type of change plus a short description. You have the choice between :
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:17]: Title tags are more often in the for `[TAG]` where the tag is minimal. I do not think that's a great idea to restrict to a list. Adding some examples is a good way of ensuring common tags are respected in a gentle way.

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+Example of title : `Bug fix : Pinneaples on pizza were removed`
+
+Your pull request need to hold following parts.
+
+## Description
+
+Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List
+any dependencies that are required for this change.
+
+## How Has This Been Tested?
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:33]: This section does not feel useful, we rely on unit tests and/or integration tests. The answer should always be "I created the relevant test cases".

**Review by galyfray — CHANGES_REQUESTED** [review]
The template should be more gentle with the reader, more "please" and other kind words, this template aims to gently guide the contributors toward a good PR. From my personal experience, the templates themselves are the skeleton of the PR, and the TODO-ish list that you wrote at the beginning would be in a "requirements" section. This way, we could monitor the fact that the contributor checked the boxes and thus should be aware that he should have completed the requirements.
Again from my experience I quickly flew away from projects where the template was aggressive mostly because when the template is aggressive the reviewers also are.

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -1,43 +1,31 @@
 ## Prerequisites
 
-***This part help you to don't forget anything before create pull request, If everything seem good you could delete it***
+***This part help you to don't forget anything before create pull request, If everything seem good you can delete it***
 
-Thanks for contribution :heart:. Before creating pull
-request you were able to check :
+Thanks for your contribution :heart:. Before creating pull
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:5]:
```suggestion
Thanks for your contribution :heart:. Before creating a pull
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 11]
```diff
@@ -1,43 +1,31 @@
...
+- [ ] You have commented your code, particularly in hard-to-understand areas
 - [ ] You have made corresponding changes to the documentation
 - [ ] Your changes generate no new warnings
-- [ ] You have added tests that prove my fix is effective or that my feature works
+- [ ] You have added tests that prove your fix is effective or that your feature works
 - [ ] New and existing unit tests pass locally with your changes
 - [ ] Any dependent changes have been merged and published in downstream modules
 
-Your pull request title must contain type of change plus a short description. You have the choice between :
+Your pull request title must contain type of change plus a short description. Title type must be relevant and minimal (one or two words), this is a non-exhaustive list:
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:17]: The tag must be relevant and minimal, the title can be longer than two words.

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -1,43 +1,31 @@
...
 
 - Bug fix (non-breaking change which fixes an issue)
 - New feature (non-breaking change which adds functionality)
 - Breaking change (fix or feature that would cause existing functionality to not work as expected)
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:21]: I am more a fan of this type of tags :

```suggestion
- [BUG]: bug fix
- [FEATURE]: changes which adds functionality
- [BREAKING]: prepend this tag to any PR that breaks retro-compatibility.
```
I prefer them as they are shorter and clearly distinct from the title.
What do you prefer ?

**Review by galyfray — CHANGES_REQUESTED** [review]
I think that the skeleton of the PR is kind of empty, we should have a more in depth discussion of what we would like a PR to looks like.

**Comment by galyfray** [comment]
We might also want peoples to add relevant closing tags to the PR that fixes bugs, see this [documentation](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) if you want more details on what I mean.

**Review by galyfray — CHANGES_REQUESTED** [review]
All I can see here is great, as said earlier I would like to discuss with you about the skeleton of the PR. Apart from this ongoing conversation, this PR is good to go !

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 13]
```diff
@@ -2,19 +2,13 @@
 
 ***This part help you to don't forget anything before create pull request, If everything seem good you can delete it***
 
-Thanks for your contribution :heart:. Before creating a pull
-request please make sure :
+Thanks for your contribution :heart:. Before creating a pull request please make sure you checked all conditions
+of [checks part](#Checks), includes all following parts and title respect rule.
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:6]:
```suggestion
of the [checks part](#Checks), including all following parts and the title rules as this makes our lives easier :heart:.
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -2,19 +2,13 @@
...
-- [ ] You have commented your code, particularly in hard-to-understand areas
-- [ ] You have made corresponding changes to the documentation
-- [ ] Your changes generate no new warnings
-- [ ] You have added tests that prove your fix is effective or that your feature works
-- [ ] New and existing unit tests pass locally with your changes
-- [ ] Any dependent changes have been merged and published in downstream modules
+### Pull request title :
 
-Your pull request title must contain a tag type plus a short description. Tag must be relevant and minimal, this is a non-exhaustive list:
+Your pull request title must contain a tag type plus a short description. Tag must be relevant and minimal, this is a
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:10]:
```suggestion
Your pull request title must contain a tag type plus a short description. Tags must be relevant and minimal, here is a
```

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 15]
```diff
@@ -23,9 +17,22 @@ Your pull request title must contain a tag type plus a short description. Tag mu
...
+## Checks
+
+- [ ] Your code follows the style guidelines of this project
+- [ ] You have performed a self-review of your code
+- [ ] You have commented your code, particularly in hard-to-understand areas
+- [ ] You have made corresponding changes to the documentation
+- [ ] Your changes generate no new warnings
+- [ ] You have added tests that prove your fix is effective or that your feature works
+- [ ] New and existing unit tests pass locally with your changes
+- [ ] Any dependent changes have been merged and published in downstream modules
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:38]: I do not think this one is really needed. I mean, the down streamed modules should update in response to the approved changes of the repo, not the opposite, right ?

**Review by galyfray — CHANGES_REQUESTED** [review]

File: PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md [hunk 16]
```diff
@@ -1,3 +1,5 @@
+Disclaimer :exclamation: Issues are formatted using markdown, a useful summary of the features provided by GitHub can be found here.
```
Comment (galyfray) [inline PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.md:1]: Missing link.

**Review by galyfray — CHANGES_REQUESTED** [review]

**Review by galyfray — CHANGES_REQUESTED** [review]
could add a part about reading the code of conduct and the contributing.md ?

**Comment by Beladric** [comment]
> could add a part about reading the code of conduct and the contributing.md ?

It's already done in checks part ^^

**Comment by galyfray** [comment]
> > could add a part about reading the code of conduct and the contributing.md ?
> 
> It's already done in checks part ^^

It seems like it is not ^^
> ```md

- [ ] Your code follows the style guidelines of this project
- [ ] You have performed a self-review of your code
- [ ] You have commented your code, particularly in hard-to-understand areas
- [ ] You have made corresponding changes to the documentation
- [ ] Your changes generate no new warnings
- [ ] You have added tests that prove your fix is effective or that your feature works
- [ ] New and existing unit tests pass locally with your changes
```

**Review by galyfray — APPROVED** [review]
All good.

**Pull request merged**